File: MaxLevelTree.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// Level order traversal
class Solution {
    
     public List<Integer> largestValues(TreeNode root) {
         List<Integer> res = new ArrayList<>();
         if(root == null)
             return res;
         
         Queue<TreeNode> q = new LinkedList<>();
         q.add(root);
         int max=0;
         
         while(!q.isEmpty()){ 
             int size = q.size();
             max = Integer.MIN_VALUE;
             for(int i=0; i<size; i++){
                 TreeNode front = q.remove();
                 
                 max = Math.max(max, front.val);
                 if(front.left != null){
                     q.add(front.left);
                 }
                 if(front.right != null){
                     q.add(front.right);
                 }
             }
             res.add(max);   
         }         
         return res;
    }
    
    
   
}

File: maxlevelTree.py
"""
Time - O(N)
Space - O(N)
Leetcode - yes
problems - no
"""

class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        
        if(root is None):
            return []
        
        result = []
        queue = []
        
        queue.append(root)
        
        while(len(queue) > 0):
            
            temp = float("-inf")
            size = len(queue)
            
            for i in range(size):
                
                # curr node
                curr = queue.pop(0)
                
                temp = max(temp, curr.val)

                if(curr.left):
                    queue.append(curr.left)
                if(curr.right):
                    queue.append(curr.right)
                            
            result.append(temp)
            
        return result
            
        
        
        